ls1intum/Artemis
# [Enhancement] Manual Result based on Automatic Result + Enable Result score > 100% for all Exercises types

### Checklist

### Motivation and Context


We want to connect to automatic feedback into the manual result, so that the tutor and student have the full range of feedback and the score is calculated overall the feedback. Additionally this PR enables results with a score > 100% for all exercise types. 

### Test Coverage

* code-editor-tutor-assessment-container.component.ts: 83.17%
* ProgrammingExerciseGradingService.java: 96%
* GradingCriterionService.java: 85%
* AssessmentService.java: 76%
* ProgrammingAssessmentService.java: 89%

### Description

- Add automatic feedback into manual result
- Set credits of automatic feedback to 0, it it does not have credits set
- Adapt check that feedbacks must contain detail text
- Set feedback to positive if it has credits > 0
- Enable test case names for result dialog for assessment view
- Enable results >100% for all exercise types
- Properly count the totalScore of all feedback elements (especially for programming exercise)
- After reevaluation after configure grading, also manual result is updated
- Change manual result AssessmentType from `MANUAL` to `SEMI_AUTOMATIC`

### Steps for Testing


FileUpload, Modeling, Text:
1. Log in to Artemis
2. Navigate to Course Administration
3. Create Text, Modeling, File Upload exercises with bonus points and participate
4. Assess the exercises and give enough points to check if the max points are capped by max score + bonus points of exercise (try also SGI)
4.1 Play around with the points you give and see if everything behaves as expected
5. Result label should display then a result > 100% (also from students view) when you gave more than the maxScore.

Programming Exercises:
1. Log in to Artemis
2. Navigate to Course Administration
3. Create programming exercises with bonus points (with SCA and one without SCA) and participate (write only `Collections.sort(input)` to `BubbleSort.java` to only solve 1 test case, for SCA you get 3 issue of category `Code Style` (Magic Number issues))
4. Assess programming exercise and click afterwards on result label to see if manual result contains automatic feedback and test case names
**NOTE**: Positive automatic result, will count into the score calculation but will not be displayed in the result dialog
4.1 Check for SCA that points are deducted, if in the category in configure grading is set to `graded` with penalties
4.2 Check that the max points are capped by max score + bonus points of exercise
4.3 Try also out SGI
5. Check also student view (students should not see test case names) 
6. Change values in configure grading and check that the automatic and manual result are updated accordingly
6.1 **NO** manual result  feedback must be lost (referenced, unreferenced and general feedback)

Examples to check if the result is calculated correctly for a exercise with maxScore=100 and bonusPoints=10:

Programming exercise without SCA but tested with SGI (one SGI is restricted to counted only 2 times, but was applied 3 times) and with Configure grading for Test Cases:
Calculation => 7,69 Points for 1 successful test case + 100 Points for successful test case + 4 Points for SGI feedback (third is not counted as SGI is restricted to 2) - 3 Points normal manual feedback = 108,69 Points

Configure Grading View:
![image](https://user-images.githubusercontent.com/41108487/97009861-ecd0d300-1544-11eb-89c5-ac92c162fc3e.png)

Assessment View:
<img width="1419" alt="image" src="https://user-images.githubusercontent.com/41108487/97011012-70d78a80-1546-11eb-96c5-82d02315c81f.png">


Programming exercise with SCA and Configure grading for SCA:
Calculation => 7,69 Points for 1 successful test case + 103 Points given per manual assessment  - 3 Points for SCA = 107,69 Points

Configure Grading View:
![image](https://user-images.githubusercontent.com/41108487/97009035-e3933680-1543-11eb-82ab-9d56f3979cb1.png)

Assessment View:
<img width="1389" alt="image" src="https://user-images.githubusercontent.com/41108487/97008944-c3637780-1543-11eb-880a-57962839fbcd.png">

### Screenshots


Positive manual feedback and test case names in assessment view:
<img width="1262" alt="image" src="https://user-images.githubusercontent.com/41108487/96379504-a4ee2c80-118f-11eb-8643-c76063e103c5.png">

Result details view from student perspective (no test case names but positive manual feedback):
<img width="1607" alt="image" src="https://user-images.githubusercontent.com/41108487/96382950-c7347a00-1190-11eb-8f04-f7712789692b.png">
